Task based refactor

markharley opened the pull request:
## Why are these changes needed?
* Refactor to facilitate improvement of time-series fitting.
* Delegate task specific logic to a Task sub-class, simplifying the AutoML class.
* Introduce a TimeSeriesDataset class for encapsulation of time-series relevant data operations.
* Introduce a TimeSeriesEstimator parent class to standardise time-series models.

## Related issue number
N/A

## Checks

- [X] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [X] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

gmdiana-hershey commented:
Was looking at the failed build and noticed that there were some import issues. Python 3.6 doesn't include data classes, but there is a backport. Adding that to the python 3.6 build script would solve that issue. Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).

int-chaos commented on flaml/ml.py:
```diff
@@ -1,710 +1,628 @@
-    for i in range(len(classes)):
-        curve_x[i], curve_y[i], _ = curve_func(y_true_binary[:, i], y_pred_proba[:, i])
-    return curve_x, curve_y
+# !
```
Entire file is shown as edited, which makes it hard to review. Make sure EOL (LF vs CRLF) is kept consistent, unless we are actually changing the EOL.

markharley replied:
Hey. Thanks for the review. Yeah, looks like one of our editors wasn't happy with CRLF. I've reverted the line endings back to DOS

int-chaos commented on flaml/default/suggest.py:
```diff
@@ -1,259 +1,260 @@
-
-        transformer = AutoMLTransformer()
-        return hyperparams, estimator_class, X, y, transformer, dt.label_transformer
+import numpy as np
```
Entire file is shown as edited, which makes it hard to review. Make sure EOL (LF vs CRLF) is kept consistent, unless we are actually changing the EOL.

markharley replied:
Hey. Thanks for the review. Yeah, looks like one of our editors wasn't happy with CRLF. I've reverted the line endings back to DOS

int-chaos commented on flaml/model.py:
```diff
@@ -1,2369 +1,1767 @@
-        # Close the null files
-        os.close(self.null_fds[0])
-        os.close(self.null_fds[1])
+# !
```
Entire file is shown as edited, which makes it hard to review. Make sure EOL (LF vs CRLF) is kept consistent, unless we are actually changing the EOL.

markharley replied:
Hey. Thanks for the review. Yeah, looks like one of our editors wasn't happy with CRLF. I've reverted the line endings back to DOS

sonichi commented on website/docs/Use-Cases/Task-Oriented-AutoML.md:
```diff
@@ -1,4 +1,4 @@
-# Task Oriented AutoML
+# GenericTask Oriented AutoML
```
Is this and other changes from Task->GenericTask in the documentation a mistake?

markharley replied:
Hey! Thanks for the review.  Yeah, this and a couple of other instances appear to be a mistaken replacement. Likely Pycharm refactor going over-the-top. I've fixed them up now

liususan091219 commented on flaml/nlp/huggingface/training_args.py:
```diff
@@ -1,9 +1,7 @@
-from ...data import (
-    NLG_TASKS,
-)
+from flaml.automl.task import NLG_TASKS
```
The `...data` is meant to be kept as the relative path to be consistent with the other files. Is there any reasons to change it to the absolute path?

liususan091219 replied:
```suggestion
from ...automl.task import NLG_TASKS
```

liususan091219 commented on flaml/nlp/huggingface/data_collator.py:
```diff
@@ -6,7 +6,11 @@
+    MULTICHOICECLASSIFICATION,
+    TOKENCLASSIFICATION,
+    SUMMARIZATION,
+)
```
Here it was meant to be updated to the relative path as well.

liususan091219 commented on flaml/nlp/huggingface/utils.py:
```diff
@@ -2,12 +2,12 @@
-from ...data import (
-    SUMMARIZATION,
-    SEQREGRESSION,
+from flaml.automl.task import (
```
The ...data is meant to be kept as the relative path to be consistent with the other files. Is there any reasons to change it to the absolute path?

liususan091219 replied:
```suggestion
from ...automl.task import (
```

liususan091219 commented on flaml/nlp/huggingface/data_collator.py:
```diff
@@ -6,7 +6,11 @@
 from collections import OrderedDict
 
-from flaml.data import TOKENCLASSIFICATION, MULTICHOICECLASSIFICATION, SUMMARIZATION
+from flaml.automl.task import (
```
```suggestion
from ...automl.task import (
```

liususan091219 commented on flaml/nlp/utils.py:
```diff
@@ -1,13 +1,12 @@
-from ..data import (
-    SUMMARIZATION,
-    SEQREGRESSION,
+from flaml.automl.task import (
```
```suggestion
from ...automl.task import (
```

liususan091219 commented on flaml/data.py:
```diff
@@ -7,47 +7,21 @@
-    "classification",
-    SEQCLASSIFICATION,
-    MULTICHOICECLASSIFICATION,
+from flaml.automl.task import (
```
```suggestion
from .automl.task import (
```

liususan091219 commented on flaml/default/estimator.py:
```diff
@@ -1,6 +1,6 @@
 import sklearn.ensemble as ensemble
 from functools import wraps
-from flaml.data import CLASSIFICATION
+from flaml.automl.task import CLASSIFICATION
```
```suggestion
from .automl.task import CLASSIFICATION
```

liususan091219 commented:
> Was looking at the failed build and noticed that there were some import issues.
> 
> * Python 3.6 doesn't include data classes, but there is a backport that could be added to the build script (and setup).
> * Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).
> * It sounds like new tests were added for time series tasks? Should the requirements in [ts_forecast] and [forecast] also be included in [test] in that case?

The holidays are only used in time series forecasting which is optional. Make sure the import statement is invoked only under the time series environment. See https://github.com/microsoft/FLAML/blob/main/flaml/automl.py#L1096 as an example for nlp (another optional environment).

qingyun-wu reviewed (commented):
Would it be more reasonable to put the "nlp" and "time_series" folders into the "automl" folder?

markharley commented:
> Was looking at the failed build and noticed that there were some import issues.
> 
> * Python 3.6 doesn't include data classes, but there is a backport that could be added to the build script (and setup).
> * Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).
> * It sounds like new tests were added for time series tasks? Should the requirements in [ts_forecast] and [forecast] also be included in [test] in that case?

Hey! Thanks for the review. I've added dataclasses as a Python version dependent install requirement, thanks for spotting it! Holidays turned out not to be used outside of the tests at present so I've removed the unnecessary test in test_forecast

markharley commented:
> > Was looking at the failed build and noticed that there were some import issues.
> > 
> > * Python 3.6 doesn't include data classes, but there is a backport that could be added to the build script (and setup).
> > * Also saw that importing holidays fails. That's not part of the [test] requirements in setup.py (see https://github.com/markharley/FLAML/blob/fafd6524c9466d8ab9c9049bb6102c34cd5819ec/setup.py#L91).
> > * It sounds like new tests were added for time series tasks? Should the requirements in [ts_forecast] and [forecast] also be included in [test] in that case?
> 
> The holidays are only used in time series forecasting which is optional. Make sure the import statement is invoked only under the time series environment. See https://github.com/microsoft/FLAML/blob/main/flaml/automl.py#L1096 as an example for nlp (another optional environment).

Hey! Thanks for the review @liususan091219. I've applied your relative import suggestions. On `holidays`, it turns out that it wasn't called outside of tests at present and so I've removed the offending test and redundant code paths

markharley commented:
> Would it be more reasonable to put the "nlp" and "time_series" folders into the "automl" folder?

Hey! I think we ended up with this layout to avoid some circular imports, but I'll have another try at refactoring these into the automl subpackage :smile:

sonichi commented:
Some checks failed. I wonder if it will be easier to break the PR down to smaller PRs. For example, the first PR to make is to just create the `automl` subpackage and move files to corresponding locations. Otherwise this PR would take a long time to merge and you will get conflicts often.

qingyun-wu commented:
> 


Thank you @markharley! We indeed need to re-organize the whole structure of the **flaml** folder considering the need of adding an automl folder. I am attaching a proposal for the new structure (considering all content in flaml, not just the changes involved in this PR). 

In this PR, perhaps you can just make .py files about automl and the time_series folder in the right place. We can come up with a plan with the other changes (and perhaps also discuss this proposed structure plan in the maintainer meeting on 10/10). 

- **automl**
     - **nlp**
     - **time_series** 
      automl.py
      data.py
      ml.py
      model.py
      train_log.py
      config.py
      [and other files need to be added in the automl folder]
- **tune**
    - **searcher**
    - **scheduler**
    [and existing files in the tune foler]
- **onlineml**
- **default**

sonichi commented on flaml/automl/task.py:
```diff
@@ -0,0 +1,184 @@
+class Task(ABC):
+    estimators = {}
+
+    def __init__(
+        self,
+        task_name: str,
+        X_train: Union[np.ndarray, pd.DataFrame],
+        y_train: Union[np.ndarray, pd.DataFrame, pd.Series],
+    ):
```
The constructor here indicates that a Task object needs to know about X_train and y_train when it's constructed. 
The implementation indicates that no reference to the dataset is stored inside the Task object.
Why? What's the relation between a Task object and a dataset exactly?

EgorKraevTransferwise replied:
for example, at this stage we could infer it it's a binary or multi-category classification, and whether it's a univariate or panel regression, so the user wouldn't have the hassle of specifying that

qingyun-wu commented on flaml/automl/task.py:
```diff
@@ -0,0 +1,184 @@
+
+    def __init__(
+        self,
+        task_name: str,
```
Needs documentation on the allowed task_name name?

qingyun-wu commented on setup.py:
```diff
@@ -20,6 +20,7 @@
     "scipy>=1.4.1",
     "pandas>=1.1.4",
     "scikit-learn>=0.24",
+    "dataclasses>=0.8 ; python_version=='3.6'",
```
```suggestion
    "dataclasses>=0.8 ; python_version>='3.6'",
```
